sdk/confidentialledger/Azure.Security.CodeTransparency/src/CodeTransparencyClient.cs

```diff
@@ -101,68 +106,122 @@ bool CertValidationCheck(X509Certificate2 cert)
+                    CcfReceiptVerifier.VerifyTransparentStatementReceipt(jsonWebKey, receipt, signedStatement);
+                }
+                catch (Exception e)
+                {
+                    failures.Add(e);
+                }
             }
-            catch (Exception e)
+            if (failures.Count > 0)
             {
-                scope.Failed(e);
-                throw;
+                throw new AggregateException(failures);
             }
         }
 
-        internal HttpMessage CreateCreateEntryRequest(RequestContent content, RequestContext context)
+        /// <summary>
+        /// Get the service certificate key from the JWKs service endpoint.
+        /// </summary>
+        /// <param name="receiptBytes">the COSE receipt bytes,
+        /// see https://www.ietf.org/archive/id/draft-ietf-cose-merkle-tree-proofs-08.html#name-verifiable-data-structures-</param>
+        /// <returns>The service certificate key (JWK)</returns>
+        private JsonWebKey GetServiceCertificateKey(byte[] receiptBytes)
         {
-            var message = _pipeline.CreateMessage(context);
-            var request = message.Request;
-            request.Method = RequestMethod.Post;
-            var uri = new RawRequestUriBuilder();
-            uri.Reset(_endpoint);
-            uri.AppendPath("/entries", false);
-            uri.AppendQuery("api-version", _apiVersion, true);
-            request.Uri = uri;
-            request.Headers.Add("Accept", "application/json");
-            request.Headers.Add("content-type", "application/cose");
-            request.Content = content;
-            return message;
+            CoseSign1Message coseSign1Message = CoseMessage.DecodeSign1(receiptBytes);
+
+            if (!coseSign1Message.ProtectedHeaders.TryGetValue(new CoseHeaderLabel(CcfReceipt.COSE_RECEIPT_CWT_MAP_LABEL), out CoseHeaderValue cwtMap))
+            {
+                throw new InvalidOperationException("CWT-MAP not found in receipt.");
+            }
+
+            string issuer = string.Empty;
+
+            CborReader cborReader = new CborReader(cwtMap.EncodedValue.ToArray());
+            cborReader.ReadStartMap();
+            while (cborReader.PeekState() != CborReaderState.EndMap)
+            {
+                int key = cborReader.ReadInt32();
+                if (key == CcfReceipt.COSE_RECEIPT_CWT_ISS_LABEL)
+                    issuer = cborReader.ReadTextString();
+                else
+                    cborReader.SkipValue();
+            }
+            cborReader.ReadEndMap();
+
+            // Validate issuer and CTS instance are matching
+            if (!issuer.Equals(_serviceName, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new InvalidOperationException("Issuer and CTS instance name are not matching.");
+            }
+
+            // Get all the public keys from the JWKS endpoint
+            JwksDocument jwksDocument = GetPublicKeys().Value;
```

Comment:
There is a possibility to get 503 here, I am not sure how to force generated code to take that into account and retry the request in that case

Reply:
Added a test to validate the retry of GetPublicKeys()